WalletWasabi/BitcoinCore/CachedRpcClient.cs

```diff
@@ -0,0 +1,209 @@
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Primitives;
+using NBitcoin;
+using NBitcoin.RPC;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WalletWasabi.BitcoinCore
+{
+	public class CachedRpcClient : RpcClientBase
+	{
+		private CancellationTokenSource _tipChangeCancellationTokenSource;
+
+		public CachedRpcClient(RPCClient rpc, IMemoryCache cache)
+			: base(rpc)
+		{
+			Cache = cache;
+		}
+
+		public IMemoryCache Cache { get; }
+
+		private CancellationTokenSource TipChangeCancellationTokenSource
+		{
+			get 
+			{
+				if (_tipChangeCancellationTokenSource is null || _tipChangeCancellationTokenSource.IsCancellationRequested)
+				{
+					_tipChangeCancellationTokenSource = new CancellationTokenSource();
```

Comment:
Can is this be created in the constructor to avoid race conditions?

Reply:
No, we cannot create one `CancellationTokenSource` because it can be cancelled anytime and then we need a new one (this cts is used to invalidate all the cache entries related to blocks). We can add a lock around the `if` block that contains the assignment. 
What do you think?

Reply:
Yes add it, that is a good idea.

Reply:
Done.